compiler/resolution/implementForallIntents2.cpp

```diff
@@ -0,0 +1,1097 @@
+  FnSymbol*    curFn;
+
+  // Calling a function nested in the iterator?
+  bool         nested;
+
+  // The iterator's return symbol.
+  VarSymbol*   retSym;
+
+  // The FnSymbol for task startup/teardown code, usually 'curFn'.
+  // It is an enclosing function when descending into a task function
+  // that does not create a user-level task. Such a task function
+  // does not affect the intents.
+  FnSymbol*    anchorFn;
+
+  // Where in 'anchorFn' to insert startup/teardown code.
+  Expr*        anchor1;
+  Expr*        anchor2;
+  Expr*        anchor2orig;
+  Expr*        anchor2prev;
+
+  // Information on each forall intent across this call.
+  FIvec        fiInfos;
+
+  bool inIterator() const { return !nested; }
+  bool inTaskFn()   const { return nested; }
+
+  int  numShadowVars()   const {
+    int result = forall->numIntentVars();
+    INT_ASSERT(result == (int)fiInfos.size());
+    return result;
+  }
+
+  // constructor for the outermodst context
+  FIcontext(ForallStmt* fs, CallExpr* iterCall,
+            FnSymbol* iterFn, VarSymbol* retSymbol) :
+    forall(fs),
+    call(iterCall),
+    curFn(iterFn),
+    nested(false),
+    retSym(retSymbol),
+    anchorFn(curFn),
+    anchor1(0), anchor2(0), anchor2orig(0), anchor2prev(0)
+  {
+    fiInfos.resize(forall->numIntentVars());
+  }
+
+  // constructor for a nested context
+  FIcontext(FIcontext& parentCx, CallExpr* taskCall, FnSymbol* taskFn) :
+    forall(parentCx.forall),
+    call(taskCall),
+    curFn(taskFn),
+    nested(true),
+    retSym(parentCx.retSym),
+    anchorFn(curFn),
+    anchor1(0), anchor2(0), anchor2orig(0), anchor2prev(0)
+  {
+    int numSVars = forall->numIntentVars();
+    fiInfos.resize(numSVars);
+    for (int ix = 0; ix < numSVars; ix++) {
+      fiInfos[ix].svSymbol = parentCx.fiInfos[ix].svSymbol;
+      fiInfos[ix].fiActual = parentCx.fiInfos[ix].fiFormal;
+    }
+  }
+};
+
+
+/////////// EFLOPI: Enclosing For Loop Over Parallel Iterator ///////////
+
+// 
+// Info we gather while looking for an eflopi or an enclosing forall loop
+//
+// The contents are valid only after a call to findEnclosingParallelLoop()
+// as follows:
+//
+//  * 'eflopiForall' and 'eflopiLoop' are always set up
+//   -  to the corresponding ForallStmt or ForLoop node if one is found,
+//   -  to NULL if neither is found.
+//
+//  * the other fields are set up only if eflopiLoop != NULL.
+//
+class EflopiInfo {
+public:
+  ForallStmt* eflopiForall;
+  ForLoop*    eflopiLoop;
+  CallExpr*   eflopiCall;
+  // more details
+  // todo soon: remove all except asgnToIter?
+  Symbol*    iterSym;
+  CallExpr*  asgnToIter;
+  Symbol*    iterCallTemp;
+  CallExpr*  asgnToCallTemp;
+};
+
+//
+// Is 'forLoop' a loop over a parallel iterator?
+// If so, fill in 'eInfo' with details.
+//
+static bool findCallToParallelIterator(ForLoop* forLoop, EflopiInfo& eInfo)
+{
+  Symbol* iterSym = forLoop->iteratorGet()->symbol();
+
+  // Find an assignment to 'iterSym'.
```

Comment:
Is this usually located just a few statements before forLoop? If so, add a comment to that effect. If not, you might use for_SymbolSymExprs.